libraries/AC_Autorotation/AC_Autorotation.cpp

```diff
@@ -2,375 +2,524 @@
-            //unhealthy, rpm unreliable
-            _flags.bad_rpm = true;
-        }
+    // Checking to ensure no nullptr, we do have a pre-arm check for this so it will be very bad if RPM has gone away
+    if (rpm == nullptr) {
+        return false;
+    }
 
-    } else {
-        _flags.bad_rpm = true;
+    // Check RPM sensor is returning a healthy status
+    if (!rpm->get_rpm(_param_rpm_instance.get(), current_rpm)) {
+        return false;
     }
-#else
-    _flags.bad_rpm = true;
 #endif
 
-    if (_flags.bad_rpm) {
-        //count unhealthy rpm updates and reset healthy rpm counter
-        _unhealthy_rpm_counter++;
-        _healthy_rpm_counter = 0;
+    // Protect against div by zeros later in the code
+    float head_speed_set_point = MAX(1.0, _param_head_speed_set_point.get());
 
-    } else if (!_flags.bad_rpm && _unhealthy_rpm_counter > 0) {
-        //poor rpm reading may have cleared.  Wait 10 update cycles to clear.
-        _healthy_rpm_counter++;
+    // Normalize the RPM by the setpoint
+    norm_rpm = current_rpm/head_speed_set_point;
+    return true;
+}
 
-        if (_healthy_rpm_counter >= 10) {
-            //poor rpm health has cleared, reset counters
-            _unhealthy_rpm_counter = 0;
-            _healthy_rpm_counter = 0;
-        }
+// Update speed controller
+void AC_Autorotation::update_forward_speed_controller(float pilot_norm_accel)
+{
+    // Limiting the desired velocity based on the max acceleration limit to get an update target
+    const float min_vel = _target_vel - get_accel_max() * _dt;
+    const float max_vel = _target_vel + get_accel_max() * _dt;
+    _target_vel = constrain_float(_desired_vel, min_vel, max_vel); // (m/s)
+
+    // Calculate acceleration target
+    const float fwd_accel_target  = _fwd_speed_pid.update_all(_target_vel, get_speed_forward(), _dt, _limit_accel); // (m/s/s)
+
+    // Build the body frame XY accel vector.
+    // Pilot can request as much as 1/2 of the max accel laterally to perform a turn.
+    // We only allow up to half as we need to prioritize building/maintaining airspeed.
+    Vector2f bf_accel_target = {fwd_accel_target, pilot_norm_accel * get_accel_max() * 0.5};
+
+    // Ensure we do not exceed the accel limit
+    _limit_accel = bf_accel_target.limit_length(get_accel_max());
+
+    // Calculate roll and pitch targets from angles, negative accel for negative pitch (pitch forward)
+    const float roll_angle_cdeg = accel_to_angle(bf_accel_target.y) * 100.0;
+    Vector2f angle_target_cdeg = { accel_to_angle(-bf_accel_target.x) * 100.0, // Pitch
+                                   roll_angle_cdeg}; // Roll
+
+    // Ensure that the requested angles do not exceed angle max
+    _limit_accel = _limit_accel || angle_target_cdeg.limit_length(_attitude_control->lean_angle_max_cd());
+
+    // we may have scaled the lateral accel in the angle limit scaling, so we need to account
+    // for the ratio in the next yaw rate calculation
```

Comment:
I think you could use `angle_to_accel` to go back to a accel again form the constrained angle rather than working out the ratio here.

Reply:
Yeah, doing this ended up cleaning up the code a bit more.  Many thanks!